app/src/main/java/com/github/polypoly/app/network/RemoteDB.kt

```diff
@@ -53,123 +47,191 @@ class RemoteDB(
         return valuePromise
     }
 
-    override fun <T : Any> getValue(key: String, clazz: KClass<T>): CompletableFuture<T> {
-        return getValueAndThen(key, { data, valuePromise ->
-            valuePromise.complete(data.getValue(clazz.java)!!)
-        }, { valuePromise ->
-            valuePromise.completeExceptionally(NoSuchElementException("No value found for key <$key>"))
-        })
-    }
+    /**
+     * First gets the [DataSnapshot] associated to the given [absoluteKey] and then composes it
+     * with the corresponding given future (onSuccess if the data exists, onFailure else)
+     * @return a future composed with [onSuccess] or [onFailure]
+     */
+    private fun <T> getDataAndThen(
+        absoluteKey: String,
+        onSuccess: (DataSnapshot) -> CompletableFuture<T>,
+        onFailure: CompletableFuture<T>
+    ): CompletableFuture<T> {
+        val waitingFuture = CompletableFuture<CompletableFuture<T>>()
 
-    override fun <T : Any> getAllValues(key: String, clazz: KClass<T>): CompletableFuture<List<T>> {
-        return getValueAndThen(key, { data, valuesPromise ->
-            val users = ArrayList<T>()
-            for (child in data.children) {
-                users.add(child.getValue(clazz.java)!!)
+        rootRef.child(absoluteKey).get().addOnSuccessListener { data ->
+            if(!data.exists()) {
+                waitingFuture.complete(onFailure)
+            } else {
+                waitingFuture.complete(onSuccess(data))
             }
```

Comment:
maybe something like this ? waitingFuture.complete(if (data.exists()) onSuccess(data) else onFailure)